### PR TITLE
feat: arithmetic generics 

### PR DESCRIPTION
**Pull request by michaeljklein**
# Description

## Problem\*

Supersedes https://github.com/noir-lang/noir/pull/4958

## Summary\*

- keeping variables up to date in the other approach was too heavy
- this approach interns the arithmetic part so that a generic arithmetic expression can be represented as:
```
(id_to_expr_with_holes, Vec<variable_in_expr>)
```

## Additional Context

Remaining issues:
- https://github.com/noir-lang/noir/issues/5149
- https://github.com/noir-lang/noir/issues/5150
- https://github.com/noir-lang/noir/issues/5151
- https://github.com/noir-lang/noir/issues/5153

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [x] **[For Experimental Features]** Documentation to be submitted in a separate PR. (https://github.com/noir-lang/noir/issues/5151)

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/node_interner.rs [hunk 1]
```diff
@@ -173,6 +176,81 @@ pub struct NodeInterner {
...
+
+// TODO: relocate
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum ArithExpr {
+    Op {
+        kind: ArithOpKind,
+        lhs: Box<ArithExpr>,
+        rhs: Box<ArithExpr>,
+    },
+    Variable(Rc<String>),
```
Comment (jfecher) [inline compiler/noirc_frontend/src/node_interner.rs:194]: Should probably keep at least the type variable id if not the whole type variable here eventually, right? To differentiate identical names if needed.

**Review by jfecher — COMMENTED** [review]
Initial approach looks good, nothing really jumped out at me

File: test_programs/execution_success/generic_arith/src/main.nr [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    // unexpected error when running these in one line:
+    //
+    // error: Failed to prove generic arithmetic equivalent:
+    // Generic arithmetic evaluated differently: Ok(2) != Err(UnboundVariable { .. })
+    // test_programs/compile_success_empty/generic_arith/src/main.nr:15:22
+    // assert(tail([1,2,3]) == [2, 3]);
+
+    // it also fails when using turbofish 
+    //
+    // assert(tail::<Field, 3>([1,2,3]) == [2, 3]);
```
Comment (vezenovm) [inline test_programs/execution_success/generic_arith/src/main.nr:115]: Does this line and the other commented out assertions also pass in the elaborator?

**Review by jfecher — CHANGES_REQUESTED** [review]
I'm hesitant on the approach here and want to hold off on merging for now.

The use of interning bloats the code a bit and it's unclear to me why we need to intern every constant used in a type position for example. I would have thought we'd only need to remember arith expressions that are unified against each other.

I also think delaying these type checks until monomorphization will probably introduce bugs that are difficult/impossible to fix without changing the approach.

I'm thinking now maybe we should just admit we're adding dependent types to Noir and try to add unification / checking for those directly. We may need to switch to a bidirectional type checker but I think it'd end up being the cleaner approach in the end and could let us eventually handle traits as well.

File: compiler/noirc_frontend/src/hir_def/types.rs [hunk 3]
```diff
@@ -908,6 +943,10 @@ impl std::fmt::Display for Type {
                 let typevars = vecmap(typevars, |var| var.id().to_string());
                 write!(f, "forall {}. {}", typevars.join(" "), typ)
             }
+            Type::GenericArith(arith_id, generics) => {
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir_def/types.rs:946]: Need a TODO: this prints the arith ID

**Comment by TomAFrench** [comment]
Closing after discussing with Michael

**Pull request closed without merging**